File: sample-rest-api/src/test/java/nl/haploid/octowight/sample/controller/CaptainControllerIT.java
package nl.haploid.octowight.sample.controller;

import nl.haploid.octowight.registry.repository.ResourceRootDmo;
import nl.haploid.octowight.registry.repository.ResourceRootDmoRepository;
import nl.haploid.octowight.sample.AbstractIT;
import nl.haploid.octowight.sample.TestData;
import nl.haploid.octowight.sample.data.CaptainResource;
import nl.haploid.octowight.sample.repository.PersonDmo;
import nl.haploid.octowight.sample.repository.PersonDmoRepository;
import nl.haploid.octowight.sample.repository.RoleDmo;
import nl.haploid.octowight.sample.repository.RoleDmoRepository;
import org.junit.After;
import org.junit.Before;
import org.junit.Test;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.MediaType;
import org.springframework.test.web.servlet.MockMvc;
import org.springframework.test.web.servlet.request.MockMvcRequestBuilders;
import org.springframework.test.web.servlet.result.MockMvcResultMatchers;
import org.springframework.test.web.servlet.setup.MockMvcBuilders;

import java.util.ArrayList;
import java.util.List;

public class CaptainControllerIT extends AbstractIT {

	private MockMvc mockMvc;

	@Autowired
	private CaptainController controller;

	@Autowired
	private PersonDmoRepository personDmoRepository;

	@Autowired
	private ResourceRootDmoRepository resourceRootDmoRepository;

	@Autowired
	private RoleDmoRepository roleDmoRepository;

	private final List<ResourceRootDmo> tempResourceRootDmos = new ArrayList<>();

	@Before
	public void setup() {
		this.mockMvc = MockMvcBuilders.standaloneSetup(controller).build();
		roleDmoRepository.deleteAllInBatch();
		personDmoRepository.deleteAllInBatch();
	}

	@Test
	public void testGetCaptain() throws Exception {
		final PersonDmo personDmo = personDmoRepository.saveAndFlush(TestData.personDmo());
		final RoleDmo roleDmo = roleDmoRepository.saveAndFlush(TestData.roleDmo(personDmo, CaptainResource.ResourceType()));
		final ResourceRootDmo resourceRootDmo = new ResourceRootDmo();
		resourceRootDmo.setAtomId(roleDmo.getId());
		resourceRootDmo.setAtomType(RoleDmo.AtomType());
		resourceRootDmo.setAtomOrigin("the seven seas");
		resourceRootDmo.setResourceId(TestData.nextLong());
		resourceRootDmo.setResourceType(CaptainResource.ResourceType());
		final ResourceRootDmo resourceRootDmoWithId = resourceRootDmoRepository.save(resourceRootDmo);
		tempResourceRootDmos.add(resourceRootDmoWithId);
		mockMvc.perform(MockMvcRequestBuilders.get(String.format("/captain/%d", resourceRootDmoWithId.getResourceId())))
				.andExpect(MockMvcResultMatchers.status().isOk())
				.andExpect(MockMvcResultMatchers.content().contentType(MediaType.APPLICATION_JSON));
	}

	@After
	public void teardown() {
		resourceRootDmoRepository.delete(tempResourceRootDmos);
	}
}

File: sample-rest-api/src/test/scala/nl/haploid/octowight/sample/controller/CaptainControllerIT.scala
package nl.haploid.octowight.sample.controller

import java.util

import nl.haploid.octowight.registry.data.Atom
import nl.haploid.octowight.registry.repository.{AtomDmo, ResourceRootDmo, ResourceRootDmoRepository}
import nl.haploid.octowight.sample.data.CaptainResource
import nl.haploid.octowight.sample.repository.{PersonDmoRepository, RoleDmo, RoleDmoRepository}
import nl.haploid.octowight.sample.{AbstractTransactionalIT, TestData}
import org.hamcrest.Matchers
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.http.MediaType
import org.springframework.test.web.servlet.MockMvc
import org.springframework.test.web.servlet.request.MockMvcRequestBuilders
import org.springframework.test.web.servlet.result.MockMvcResultMatchers
import org.springframework.test.web.servlet.setup.MockMvcBuilders
import org.springframework.web.context.WebApplicationContext

class CaptainControllerIT extends AbstractTransactionalIT {
  @Autowired private[this] val personDmoRepository: PersonDmoRepository = null
  @Autowired private[this] val resourceRootDmoRepository: ResourceRootDmoRepository = null
  @Autowired private[this] val roleDmoRepository: RoleDmoRepository = null
  @Autowired private[this] val webApplicationContext: WebApplicationContext = null

  private[this] val tempResourceRootDmos = new util.ArrayList[ResourceRootDmo]
  private[this] var mockMvc: MockMvc = null

  behavior of "Captain controller"

  override def beforeEach() = {
    super.beforeEach()
    this.mockMvc = MockMvcBuilders.webAppContextSetup(webApplicationContext).build
    roleDmoRepository.deleteAllInBatch()
    personDmoRepository.deleteAllInBatch()
  }

  override def afterEach() = {
    resourceRootDmoRepository.delete(tempResourceRootDmos)
    super.afterEach()
  }

  it should "get a captain" in {
    val personDmo = personDmoRepository.saveAndFlush(TestData.personDmo)
    val roleDmo = roleDmoRepository.saveAndFlush(TestData.roleDmo(personDmo, CaptainResource.ResourceCollection))
    val resourceRootDmo = new ResourceRootDmo(
      root = AtomDmo(new Atom(roleDmo.getId, RoleDmo.AtomCategory, "the seven seas")),
      resourceId = TestData.nextLong,
      resourceCollection = CaptainResource.ResourceCollection,
      version = TestData.nextLong)
    val resourceRootDmoWithId = resourceRootDmoRepository.save(resourceRootDmo)
    tempResourceRootDmos.add(resourceRootDmoWithId)
    mockMvc.perform(MockMvcRequestBuilders.get(s"/captain/${resourceRootDmoWithId.resourceId}"))
      .andExpect(MockMvcResultMatchers.status.isOk)
      .andExpect(MockMvcResultMatchers.content.contentType(MediaType.APPLICATION_JSON))
      .andExpect(MockMvcResultMatchers.content().string(Matchers.containsString(personDmo.getName)))
  }
}
